src/main/java/ch/cern/eam/wshub/core/services/administration/impl/UserGroupMenuServiceImpl.java

```diff
@@ -0,0 +1,321 @@
+        if ((words.length - existingPath.length) > 0) { // Path is incomplete (or doesn't exist), since differs more than 0
+            UserGroupMenuService.MenuSpecification recursiveMenuSpecification = new UserGroupMenuService.MenuSpecification(node.path, "", node.userGroup); // If function is there, then we remove only that
+            if (node.menuCode.isEmpty()) { // But if no function, then we remove only one path item at the end
+                recursiveMenuSpecification = new UserGroupMenuService.MenuSpecification(String.join("/", Arrays.copyOf(words, words.length - 1)), node.menuCode, node.userGroup);
+            }
+            this.addToMenuHierarchy(context, recursiveMenuSpecification);
+            menuEntries = this.getExtMenuHierarchyAsList(context, recursiveMenuSpecification); //TODO Calling a get two times is not good; fixed by using a HashMap or a tree in a future iteration
+        }
+
+        // Now add leaf item
+        // Decide menu type: if function type is not set, then assume it's a menu and not a function
+        String menuType = this.decideMenuType(node.menuCode, words);
+
+        // Find parent id of new entry/item from previous list
+        GenericMenuEntry parent;
+        if (menuType.equals("M") || menuType.equals("F")) { //TODO better with an enum, even if it's not our design (but might be too cluttered)
+            parent = this.getEntryByPathFromList(menuEntries, Arrays.copyOf(words, words.length - 1));
+        } else { // If to be added is a function, get end of path
+            parent = this.getEntryByPathFromList(menuEntries, words);
+        }
+
+        // With the previous ID found and the menu type determined, fill the request object for both menu item or function item
+        String id = parent != null ? parent.getId() : ""; // Which would be the extMenuCode of the parent..
+        MP6043_AddExtMenus_001 addExtMenus = new MP6043_AddExtMenus_001();
+        ExtMenus extMenus = new ExtMenus();
+        addExtMenus.setExtMenus(extMenus);
+        extMenus.setUSERGROUPID(new USERGROUPID_Type());
+        extMenus.getUSERGROUPID().setUSERGROUPCODE(node.userGroup);
+        extMenus.setFUNCTIONID(new FUNCTIONID_Type());
+        String menuCodeToSend = node.menuCode;
+        if (menuType.equals("M") || menuType.equals("F")) { //TODO better with an enum, even if it's not our design (but might be too cluttered)
+            menuCodeToSend = "BSFOLD"; // And set internal BSFOLD code for menu item
+            extMenus.getFUNCTIONID().setFUNCTIONDESCRIPTION(words[words.length - 1]); // The last item on the path provided (the name of the menu item to add)
+        }
+        extMenus.getFUNCTIONID().setFUNCTIONCODE(menuCodeToSend); // Menu code is function code..
+        extMenus.setEXTMENUPARENT(id);
+        extMenus.setEXTMENUTYPE(menuType);
+        extMenus.setSEQUENCENUMBER(100); //TODO set to a different number depending on position
```

Comment:
This is a //TODO that needs to be solved before merge

Reply:
Solved.